### ISSUE TITLE
Idea: databinding

### ISSUE THREAD

**Issue by Buggaboo** [tobykurien/Xtendroid#89]
Hi,

I'm gonna work on [data-binding](http://developer.android.com/tools/data-binding/guide.html).

**Comment by tobykurien** [comment]
Cool stuff!

**Comment by Buggaboo** [comment]
After reading up on different binding mechanisms (e.g. one-way, two-way, threesome), I decided to do it in a way inspired by [ButterKnife](http://jakewharton.github.io/butterknife/).
I like the ViewModel concept tho.

WiP.

Arguments against binding the android way:
* [Do I really need a DSL in my view code?](http://catinean.com/2015/05/31/how-you-can-go-wrong-with-the-new-data-binding-api/)

**Comment by tobykurien** [comment]
I agree about an annotation for Observables, although it doesn't have to be specific to this data binding. I have a more sophisticated observable in mind that I used for displaying data from a bluetooth device (where data was streaming in at 1Mbps): 

- you can register listeners for property changes and it will call back with which property changed
- in addition, if you are still busy processing a property change in the UI thread when more property changes happen, it discards the changes for the property you're busy with, and queues up the changes for the other properties.

To me, with Xtendroid's easy access to UI widgets, I find it easier to use than the Android data binding library (even for 2-way binding). In both cases, you still have to register your listeners, TextWatchers, etc. I prefer doing this in code than in the XML file.

**Comment by Buggaboo** [comment]
Hi,

I tried setting it up under Xtendroid, with a really simple example. I can generate the `MainActivityBinding.java` helper that sets uses the generated `BR.java`, with android-gradle.

This is all nice, but when I start referring to these via Xtend, then it complains that these types don't exist, i.e. Xtend can't find them.

[Buggaboo/Xtendroid/v0.13_data_binding](https://github.com/Buggaboo/Xtendroid/tree/v0.13_data_binding)

I need help with this one...

**Comment by Buggaboo** [comment]
No changes are required for the gradle files to build [this](https://github.com/Buggaboo/Xtendroid/tree/v0.13_data_binding). Just patience.

1) Transpile xtend without referring to the generated java files.
i.e. out comment ...DataBindingLayout...Binding.java in the xtend code,
e.g. so xtend won't fail due to the missing yet-to-be-generated
java files.
2) build with: ./gradlew :Quotes:build
3) Make the out commented lines active again. Build.
4) build again with ./gradlew :Quotes:build
For every build, you need to repeat this process.
Yes, this is a pain in the ass.

**Comment by oehme** [comment]
@Buggaboo Sorry for my ignorance about Android features :) Is databinding something that comes with the android build tools? If so, we should make the xtext-gradle-plugin aware of it.

**Comment by Buggaboo** [comment]
Yes, `com.android.tools.build:gradle` supports it from 1.5 onwards.

```gradle
android {
    dataBinding {
        enabled = true
    }
}
```
Right now we don't use the xtext-gradle-plugin for building. We use the old one (xtend-gradle-plugin) because, I can use the multi-project trick, see this [build.gradle](https://github.com/tobykurien/Xtendroid/blob/v0.13_development/build.gradle).

**Comment by oehme** [comment]
Sorry I didn’t get why you use the old one, they should be functionally equivalent.

**Comment by Buggaboo** [comment]
This is why: [xtext/xtend plugin error](https://github.com/Buggaboo/Xtendroid/blob/xtext-gradle-plugin/xtext-gradle-plugin-error.txt)

Please use this [branch](https://github.com/Buggaboo/Xtendroid/tree/xtext-gradle-plugin) to test.

**Comment by oehme** [comment]
Well first of all if you want to depend on another project, you need to use `compile project(':projectName‘)`. Because what you currently have (`compile ':someName‘`) is an external dependency. And since your buildscript does not define any repositories, these cannot be resolved. That is unrelated to the Xtend plugin.

**Comment by Buggaboo** [comment]
Ah. Excellent. Okay, I tried `compile project(':Project')`. It works. One problem less. Thanks!

Now, for databinding...

**Comment by oehme** [comment]
For databinding, please open an issue at the xtext-gradle-plugin on Github =)

**Comment by Buggaboo** [comment]
After some manual debugging with gradle, I found out that there is a cyclic dependency between:
* `"compile${variantName}JavaWithJavac"`, this creates the `*Binding.java` type required for data binding
* `"generate${variantName}Xtext"` runs before the previous task

So somehow `"compile${variantName}JavaWithJavac"` must be run twice, before `"generate${variantName}Xtext"`.
